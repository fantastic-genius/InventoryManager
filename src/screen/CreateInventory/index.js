import React, {useState, useRef} from 'react';
import {
  View,
  Text,
  TextInput,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
} from 'react-native';
import {Formik} from 'formik';
import * as Yup from 'yup';
import {CommonActions} from '@react-navigation/native';

//constants
import colors from '../../constants/colors';

//crud
import {addInventory, inventoryNameExist} from '../../crud';

const CreateInventory = ({navigation}) => {
  const formikRef = useRef();
  const descriptionRegex = /^\S+\s+\S+\s+\S+/;

  const validationSchema = Yup.object().shape({
    name: Yup.string()
      .trim()
      .required('Name is required'),
    description: Yup.string()
      .trim()
      .required('description is required')
      .matches(descriptionRegex, 'description must have at least three words'),
    price: Yup.number()
      .positive('price must be a positive value')
      .required('price is required'),
    stock: Yup.number()
      .positive('stock must be a positive value')
      .required('stock is required'),
  });

  return (
    <View style={styles.container}>
      <ScrollView style={styles.scrollCon}>
        <Formik
          initialValues={{
            name: '',
            description: '',
            price: '',
            stock: '',
          }}
          onSubmit={async (values, {setSubmitting}) => {
            const nameExists = await inventoryNameExist(values.name);
            if (nameExists) {
              if (formikRef.current) {
                formikRef.current.setFieldError('name', 'Name must be unique');
              }
              return;
            }
            await addInventory(values);
            setSubmitting(false);
            navigation.dispatch(
              CommonActions.reset({
                index: 0,
                routes: [{name: 'Home'}],
              }),
            );
          }}
          validationSchema={validationSchema}
          innerRef={formikRef}
        >
          {({
            handleChange,
            handleSubmit,
            values,
            errors,
            touched,
            isValid,
            handleBlur,
          }) => (
            <View style={styles.formCon}>
              <View style={styles.inputCon}>
                <TextInput
                  name="name"
                  style={styles.input}
                  placeholder="Name"
                  placeholderTextColor={colors.LIGHTGRAY}
                  value={values.name}
                  onChangeText={handleChange('name')}
                  onBlur={handleBlur('name')}
                  maxLength={30}
                />
                {errors.name && touched.name ? (
                  <Text style={styles.errorText}>{errors.name}</Text>
                ) : null}
              </View>
              <View style={styles.inputCon}>
                <TextInput
                  name="description"
                  style={styles.input}
                  placeholder="Description"
                  placeholderTextColor={colors.LIGHTGRAY}
                  value={values.description}
                  onChangeText={handleChange('description')}
                  onBlur={handleBlur('description')}
                  multiline
                  numberOfLines={3}
                  maxLength={100}
                />
                {errors.description && touched.description ? (
                  <Text style={styles.errorText}>{errors.description}</Text>
                ) : null}
              </View>
              <View style={styles.inputCon}>
                <TextInput
                  name="price"
                  style={styles.input}
                  placeholder="Price"
                  placeholderTextColor={colors.LIGHTGRAY}
                  value={values.price}
                  onChangeText={handleChange('price')}
                  onBlur={handleBlur('price')}
                />
                {errors.price && touched.price ? (
                  <Text style={styles.errorText}>{errors.price}</Text>
                ) : null}
              </View>
              <View style={styles.inputCon}>
                <TextInput
                  name="stock"
                  style={styles.input}
                  placeholder="Stock"
                  placeholderTextColor={colors.LIGHTGRAY}
                  value={values.stock}
                  onChangeText={handleChange('stock')}
                  onBlur={handleBlur('stock')}
                />
                {errors.stock && touched.stock ? (
                  <Text style={styles.errorText}>{errors.stock}</Text>
                ) : null}
              </View>
              <TouchableOpacity
                style={styles.addBtn}
                onPress={handleSubmit}
                disabled={!isValid}
              >
                <Text style={styles.btnText}>Add</Text>
              </TouchableOpacity>
            </View>
          )}
        </Formik>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.BACKGROUND,
  },
  scrollCon: {
    flexGrow: 1,
    padding: 16,
  },
  input: {
    flex: 1,
    color: colors.MEDIUMGRAY,
    paddingHorizontal: 10,
    backgroundColor: colors.WHITE,
    fontSize: 16,
    borderRadius: 10,
  },
  inputCon: {
    marginBottom: 16,
  },
  formCon: {
    paddingVertical: 20,
  },
  addBtn: {
    paddingHorizontal: 40,
    paddingVertical: 16,
    backgroundColor: colors.PRIMARY,
    borderRadius: 10,
    marginTop: 24,
    width: '50%',
    alignSelf: 'center',
  },
  btnText: {
    fontSize: 16,
    color: colors.WHITE,
    textAlign: 'center',
  },
  errorText: {
    fontSize: 14,
    color: colors.DANGER,
    marginTop: 5,
  },
});

export default CreateInventory;
