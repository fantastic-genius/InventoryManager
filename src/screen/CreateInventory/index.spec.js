import React from 'react';
import renderer from 'react-test-renderer';
import CreateInventory from './index';

describe('CreateInventory', () => {
  it('renders screen as expected', async () => {
    const component = renderer.create(<CreateInventory />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
