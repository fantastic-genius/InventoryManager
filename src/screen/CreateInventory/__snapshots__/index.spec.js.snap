// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateInventory renders screen as expected 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "flexGrow": 1,
        "padding": 16,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "paddingVertical": 20,
          }
        }
      >
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            maxLength={30}
            name="name"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Name"
            placeholderTextColor="#718596"
            rejectResponderTermination={true}
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "color": "#627483",
                "flex": 1,
                "fontSize": 16,
                "paddingHorizontal": 10,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            maxLength={100}
            multiline={true}
            name="description"
            numberOfLines={3}
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Description"
            placeholderTextColor="#718596"
            rejectResponderTermination={true}
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "color": "#627483",
                "flex": 1,
                "fontSize": 16,
                "paddingHorizontal": 10,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            name="price"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Price"
            placeholderTextColor="#718596"
            rejectResponderTermination={true}
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "color": "#627483",
                "flex": 1,
                "fontSize": 16,
                "paddingHorizontal": 10,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            name="stock"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Stock"
            placeholderTextColor="#718596"
            rejectResponderTermination={true}
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "color": "#627483",
                "flex": 1,
                "fontSize": 16,
                "paddingHorizontal": 10,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "#7E42F5",
              "borderRadius": 10,
              "marginTop": 24,
              "opacity": 1,
              "paddingHorizontal": 40,
              "paddingVertical": 16,
              "width": "50%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 16,
                "textAlign": "center",
              }
            }
          >
            Add
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
